### PR TITLE
[Upgrade Ubuntu] uk_staging2

### LINKED ISSUE #938: [Upgrade Ubuntu] uk_staging
- [x] setup
- [x] provision
- [x] deploy
- [x] ofn-security: provision
- [x] db_transfer
- [x] DNS switchover
   - [x] ofn-install:provision with new url
   - [x] systemctl restart puma
- connect metabase:
  - [x] `db_integrations`
  - [x] Trigger sync: https://data.openfoodnetwork.org.uk/admin/databases/88
         hmmm... permission still missing?
-  Check:
   - [x] New Relic
   - [x] Downtime alerts (new relic ~~and/or wormly?~~)
- Hook up GHA deploy
  - [x] ofn-install:setup_remote_deployment (aka setup_semaphore_deployment)
  - [x] authorise 'semaphore' deployment key for user `ofn-deploy` (see https://github.com/openfoodfoundation/ofn-install/issues/941#issuecomment-2373076883 )
  - [x] Add known_host entry for new server to GH secret DEPLOYMENT_HOSTS
  - [x] check [Deploy to staging](https://github.com/openfoodfoundation/openfoodnetwork/actions/workflows/stage.yml) works.
- Update documentation: 
   * [x] https://github.com/openfoodfoundation/ofn-install/wiki/Current-servers#uk-1 (inlude note about redis version pinned)
   * [x] document about provisioning staging servers with remote deployment (refer https://github.com/openfoodfoundation/ofn-install/issues/941)
   * [ ] anything else

### PR DESCRIPTION
**Pull request by dacook**
- First server for  #157
- Closes #938 

```
ansible-playbook playbooks/fetch_secrets.yml
ansible-playbook site.yml --limit=uk_staging2  -e "ansible_user=ubuntu"
```

### Documentation
Updated: 
* [x] https://github.com/openfoodfoundation/ofn-install/wiki/Provisioning
* [x] https://github.com/openfoodfoundation/ofn-install/wiki/Current-servers#uk-1 (inlude note about redis version pinned)
* [x] https://github.com/openfoodfoundation/ofn-install/wiki/Set-up-remote-deployment-for-new-server-%28GitHub-Actions%29

### CODE REVIEW

**Comment by dacook** [comment]
Currently unable to generate certbot cert:

```
   Domain: staging2.openfoodnetwork.org.uk
   Type:   connection
   Detail: 193.108.54.81: Fetching
   http://staging2.openfoodnetwork.org.uk/.well-known/acme-challenge/HlO5FDrdwTwwwW7D0jB5AmxTz5XYUoMyuJPzis0elyQ:
   Timeout during connect (likely firewall problem)

   To fix these errors, please make sure that your domain name was
   entered correctly and the DNS A/AAAA record(s) for that domain
   contain(s) the right IP address. Additionally, please check that
   your computer has a publicly routable IP address and that no
   firewalls are preventing the server from communicating with the
   client. If you're using the webroot plugin, you should also verify
   that you are serving files from the webroot path you provided.
```

Even though:
```
ubuntu@ov-a4c948:/etc/nginx$ sudo ufw status
Status: inactive

ubuntu@ov-a4c948:/etc/nginx$ sudo iptables -A INPUT -p tcp --dport 80 -j ACCEPT
ubuntu@ov-a4c948:/etc/nginx/sites-enabled$ sudo iptables -L
Chain INPUT (policy ACCEPT)
target     prot opt source               destination
ACCEPT     tcp  --  anywhere             anywhere             tcp dpt:http
```

**Comment by dacook** [comment]
This was due to the Infomaniak firewall. That's been disabled and I've now completed provisioning.

However there's a problem with the Redis version. Sidekiq says:

```
ofn-admin@ov-a4c948:/home/openfoodnetwork/apps/openfoodnetwork/current/log$ grep redis sidekiq.log
2024-09-12T06:38:33.729Z pid=148414 tid=3b1m INFO: Sidekiq 7.2.4 connecting to Redis with options {:size=>10, :pool_name=>"internal", :url=>"redis://localhost:6381/0", :network_timeout=>5}
redis-client requires Redis 6+ with HELLO command available (redis://localhost:6381)
```
```
ofn-admin@ov-a4c948:/$ sudo apt-get install redis-server
...
redis-server is already the newest version (5:5.0.7-2ubuntu0.1+esm2).
```
I'm a bit confused about where to install redis from. The galaxy role we use installs it with `ppa:redislabs/redis` (https://github.com/libre-ops/multi_redis/blob/main/tasks/install.yml).
But [that package says](https://launchpad.net/~redislabs/+archive/ubuntu/redis):
> Redis now has an official APT repository. We urge users to use the APT repository, as described here:
https://redis.io/docs/getting-started/installation/install-redis-on-linux/

Those installation steps involve downloading stuff with curl, which I'm unsure about.

[Another website says](https://docs.vultr.com/install-and-configure-redis-on-ubuntu-20-04) to use `ppa:chris-lea/redis-server`, which seems to load redis `6.0.5`.

**Comment by mkllnk** [comment]
I looked at fr-prod to see how that is working. It has Redis 7, installed via apt.

```
$ apt list redis-server -a
Listing... Done
redis-server/focal 6:7.4.0-1rl1~focal1 amd64 [upgradable from: 6:7.0.5-1rl1~focal1]
redis-server/now 6:7.0.5-1rl1~focal1 amd64 [installed,upgradable to: 6:7.4.0-1rl1~focal1]
redis-server/focal-updates,focal-security 5:5.0.7-2ubuntu0.1 amd64
redis-server/focal 5:5.0.7-2 amd64

$ grep redis /etc/apt/ -r
/etc/apt/sources.list.d/ppa_redislabs_redis_focal.list:deb http://ppa.launchpad.net/redislabs/redis/ubuntu focal main
```

That PPA is not on UK staging2. Reading ofn-install and Matt's multi_redis role, it should be there. Did it get skipped somehow?

**Comment by dacook** [comment]
> That PPA is not on UK staging2. Reading ofn-install and Matt's multi_redis role, it should be there. Did it get skipped somehow?

Actually I removed it, to see if that helped get the right version (but it didn't). 
Thanks for looking at fr-prod. Honestly I don't think I understand how packages work..

**Comment by dacook** [comment]
So I re-added the `ppa:redislabs/redis` PPA with the `multi_redis` role. I can see v6 in apt, but for some reason it's not being selected:

```
$ apt-cache policy redis-server
redis-server:
  Installed: 5:5.0.7-2ubuntu0.1+esm2
  Candidate: 5:5.0.7-2ubuntu0.1+esm2
  Version table:
     6:7.4.0-1rl1~focal1 500
        500 http://ppa.launchpad.net/redislabs/redis/ubuntu focal/main amd64 Packages
 *** 5:5.0.7-2ubuntu0.1+esm2 510
        510 https://esm.ubuntu.com/apps/ubuntu focal-apps-security/main amd64 Packages
        100 /var/lib/dpkg/status
...
```

Ok that's why it's not being selected:
```
$ sudo apt-get install redis-server=6:7.4.0-1rl1~focal1
...
The following packages have unmet dependencies:
 redis-server : Depends: redis-tools (= 6:7.4.0-1rl1~focal1) but 5:5.0.7-2ubuntu0.1+esm2 is to be installed
E: Unable to correct problems, you have held broken packages.
```

So.. I manually installed that dependency version too, then `redis-server`, then `redis` for good measure:
```
$ sudo apt-get install redis-tools=6:7.4.0-1rl1~focal1
...
$ sudo apt-get install redis-server=6:7.4.0-1rl1~focal1
...
$ sudo apt-get install redis=6:7.4.0-1rl1~focal1
```

Yay, now I have the right version installed, it should work, right?
```
2024-09-18T07:44:11.381Z pid=1636568 tid=yy8c INFO: Sidekiq 7.2.4 connecting to Redis with options {:size=>10, :pool_name=>"internal", :url=>"redis://localhost:6381/0", :network_timeout=>5}
redis-client requires Redis 6+ with HELLO command available (redis://localhost:6381)
```
🫠. Maybe I should ensure redis has restarted? No, it is not running and is "masked"

```
$ tail -n1 /var/log/redis/redis-server.log
1614772:M 18 Sep 2024 07:41:03.867 # Redis is now ready to exit, bye bye...

$ sudo systemctl restart redis-server
Failed to restart redis-server.service: Unit redis-server.service is masked.
```
Apparently this means it's more disabled than `disabled`, and probably means that the service shouldn't be started. I don't think I know enough to be un-masking it, so will ask for help..

**Comment by mkllnk** [comment]
I'm not quite sure what's going on here either. I tried to unmask the server but afterwards it's still masked. I don't know why.

So to start from a clean slate, I uninstalled everything:

```
apt purge redis*
```

The purge makes sure that all config is removed as well. Maybe there was some conflict? It didn't ask me about data removal. So not sure what it did there.

Trying to install redis again, it wants to install the old version. Looking at the policy, the security updates repository seems to have priority. I think that `510` means it's higher than `500` and therefore selected.

```
# apt-cache policy redis-server
redis-server:
  Installed: (none)
  Candidate: 5:5.0.7-2ubuntu0.1+esm2
  Version table:
     6:7.4.0-1rl1~focal1 500
        500 http://ppa.launchpad.net/redislabs/redis/ubuntu focal/main amd64 Packages
     5:5.0.7-2ubuntu0.1+esm2 510
        510 https://esm.ubuntu.com/apps/ubuntu focal-apps-security/main amd64 Packages
     5:5.0.7-2ubuntu0.1 500
        500 http://az2.clouds.archive.ubuntu.com/ubuntu focal-updates/universe amd64 Packages
        500 http://security.ubuntu.com/ubuntu focal-security/universe amd64 Packages
     5:5.0.7-2 500
        500 http://az2.clouds.archive.ubuntu.com/ubuntu focal/universe amd64 Packages
```

That may come from `/etc/apt/preferences.d/ubuntu-pro-esm-apps`. I'm noticing the _pro_. I think that Garethe added his Ubuntu Pro subscription which includes additional security updates. Maybe it's so safe to mask a service that is not the latest security version?

So I looked into pinning and set up a preference file at `/etc/apt/preferences.d/redislabs-ubuntu-redis-focal`:
```
Package: *
Pin: release o=LP-PPA-redislabs-redis
Pin-Priority: 511
```

Then `apt install redis-server` installs the right version. The service is active. :tada:

**Comment by dacook** [comment]
Successful deployment now, after restarting and re-provisioning with the multi_redis role.
```
ansible-playbook playbooks/deploy.yml --limit=uk_staging2
...
staging2.openfoodnetwork.org.uk : ok=31   changed=18   unreachable=0    failed=0    skipped=7    rescued=0    ignored=0
```

Now to provision ofn-security. It went so well.. until right at the end:
```
~/projects/ofn-security $ ansible-playbook playbooks/provision.yml --limit uk_staging2
...
RUNNING HANDLER [suricata : reload suricata rules] *********************************************************************
fatal: [staging2.openfoodnetwork.org.uk]: FAILED! => {"changed": true, "cmd": ["pkill", "-USR2", "--pidfile", "/var/run/suricata.pid"], "delta": "0:00:00.010499", "end": "2024-09-19 02:03:09.584525", "msg": "non-zero return code", "rc": 1, "start": "2024-09-19 02:03:09.574026", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
```

It was left in "exited" state which doesn't sound good. So I manually restarted (which is the [next step](https://github.com/openfoodfoundation/ofn-security/blob/master/roles/suricata/handlers/main.yml#L11-L15) anyway)
```
ofn-admin@ov-a4c948:~$ systemctl status suricata
● suricata.service - LSB: Next Generation IDS/IPS
     Loaded: loaded (/etc/init.d/suricata; generated)
     Active: active (exited) since Thu 2024-09-19 02:00:06 UTC; 1h 9min ago
...
ofn-admin@ov-a4c948:~$ systemctl status suricata
● suricata.service - LSB: Next Generation IDS/IPS
     Loaded: loaded (/etc/init.d/suricata; generated)
     Active: active (running) since Thu 2024-09-19 03:11:44 UTC; 5s ago

ofn-admin@ov-a4c948:~$ tail -n1 /var/log/suricata/suricata.log
[47922 - Suricata-Main] 2024-09-19 03:11:57 Notice: threads: Threads created -> W: 2 FM: 1 FR: 1   Engine started.
```

🤷 I'm not sure exactly what the "reload suricate rules" task is trying to do, but if suricata is functioning then I don't want to investigate!

**Comment by dacook** [comment]
The [host_id](https://github.com/openfoodfoundation/ofn-install/blob/master/inventory/host_vars/staging.openfoodnetwork.org.uk/config.yml#L4) variable is only used for the new relic data. I'd like to :
- rename to match the new ansible host names (with underscore instead of dash)
- add a provision task to set the server hostname with it (it would be really handy to help make clear which server you're connected to)

**Comment by dacook** [comment]
Have a couple of weird errors when trying deploy.yml. 

## From my machine:

```
TASK [deploy : run yarn install] ***************************************************************************************
changed: [staging.openfoodnetwork.org.uk]
...
TASK [deploy : precompile assets] **************************************************************************************
fatal: [staging.openfoodnetwork.org.uk]: FAILED! => {
...
Yarn not installed. Please download and install Yarn from https://yarnpkg.com/lang/en/docs/install/
```
Indeed it is not, nor is nodenv:
```
openfoodnetwork@uk-staging:~/apps/openfoodnetwork/current$ yarn --version

Command 'yarn' not found, but can be installed with:
...
openfoodnetwork@uk-staging:~/apps/openfoodnetwork/current$ nodenv

Command 'nodenv' not found, did you mean:
```


## From GitHub action via ofn-deploy user:
```
TASK [deploy : gather host services facts] *************************************
fatal: [staging.openfoodnetwork.org.uk]: FAILED! => {"changed": false, "msg": "Malformed output discovered from systemd list-unit-files: accounts-daemon.service                               enabled         enabled      "}
```

**Comment by mkllnk** [comment]
I don't know why it was missing but the .bash_profile file didn't have the nodenv setup as defined in the playbook. I added that to get a bit further.

**Comment by dacook** [comment]
Ah ha, thank you! I can see that we set up the bash_profile in two steps in the provision playbook. 

First at https://github.com/openfoodfoundation/ofn-install/blob/master/roles/app_user/tasks/main.yml#L46-L53
Then at https://github.com/openfoodfoundation/ofn-install/blob/master/roles/node/tasks/main.yml#L18-L24

I executed this yesterday afternoon and it failed halfway through. I guess I didn't retry it..
We could potentially update the playbook to avoid clobbering it in the first step, but probably not work fixing right now.

**Comment by dacook** [comment]
For reference, here's the error occurring for ansible built-in command `service_facts`:
`fatal: [staging.openfoodnetwork.org.uk]: FAILED! => {"changed": false, "msg": "Malformed output discovered from systemd list-unit-files: accounts-daemon.service                               enabled         enabled      "}`

**Comment by dacook** [comment]
Although I'm not quite finished with uk_staging server, I think it's high time this PR is reviewed.
If further changes are required for metabase, I'll raise a new PR.

File: roles/deploy/tasks/deploy.yml [hunk 1]
```diff
@@ -11,9 +11,6 @@
     force: yes
   tags: clone
 
-- name: run preflight checks
-  include_tasks: preflight_checks.yml
```
Comment (mkllnk) [inline roles/deploy/tasks/deploy.yml:15]: Good call.

File: inventory/group_vars/all.yml [hunk 2]
```diff
@@ -50,18 +50,15 @@ swapfile_size: false
...
 ruby_compile_options: "--with-jemalloc"
 
 ruby_versions:
-  - version: 2.7.3
-    env:
-      RUBY_CONFIGURE_OPTS: "{{ ruby_compile_options }}"
-  - version: 3.0.3
+  - version: 3.1.4
     env:
       RUBY_CONFIGURE_OPTS: "{{ ruby_compile_options }}"
```
Comment (mkllnk) [inline inventory/group_vars/all.yml:63]: Ah, good one. New Ruby versions are now automatically installed during deploys. So we don't need to run Ansible for that. This is just for the initial version.

I wish that we could skip the version and take the version number from the ofn code but this role requires a version. So if we want to set up rbenv without installing ruby straight away then we can't use this role. Just one of the annoying Ansible things that we could replace with shell scripts...

**Review by mkllnk — APPROVED** [review]
Good work.

**Review by rioug — APPROVED** [review]
Nice one !

**Pull request merged**